SpatialGDK/Source/SpatialGDK/Private/LoadBalancing/LayeredLBStrategy.cpp

```diff
@@ -0,0 +1,276 @@
+// 					continue;
+// 				}
+// 
+// 				UAbstractLBStrategy* LBStrategy = NewObject<UAbstractLBStrategy>(this, WorkerLBLayers[Layer.Key].LoadBalanceStrategy);
+// 				LayerNameToLBStrategy.Add(LayerKey, LBStrategy);
+// 				LBStrategy->Init();
+// 				LBStrategy->AddToRoot();
+// 
+// 				for (const TSoftClassPtr<AActor>& ClassPtr : Layer.Value.ActorClasses)
+// 				{
+// 					ClassPathToLayer.Add(ClassPtr, LayerKey);
+// 				}
+// 			}
+
+			// Finally, add the default layer.
+			if (WorldSettings->LoadBalanceStrategy == nullptr)
+			{
+				UE_LOG(LogLayeredLBStrategy, Error, TEXT("If EnableUnrealLoadBalancer is set, there must be a LoadBalancing strategy set. Using a 1x1 grid."));
+				UAbstractLBStrategy* DefaultLBStrategy = NewObject<UGridBasedLBStrategy>(this);
+				DefaultLBStrategy->AddToRoot();
+				LayerNameToLBStrategy.Add(SpatialConstants::DefaultLayer, DefaultLBStrategy);
+				LayerNameToLBStrategy[SpatialConstants::DefaultLayer]->Init();
+			}
+			else
+			{
+				UAbstractLBStrategy* DefaultLBStrategy = NewObject<UAbstractLBStrategy>(this, WorldSettings->LoadBalanceStrategy);
+				DefaultLBStrategy->AddToRoot();
+				LayerNameToLBStrategy.Add(SpatialConstants::DefaultLayer, DefaultLBStrategy);
+				LayerNameToLBStrategy[SpatialConstants::DefaultLayer]->Init();
+			}
+		}
+	}
+}
+
+void ULayeredLBStrategy::SetLocalVirtualWorkerId(VirtualWorkerId InLocalVirtualWorkerId)
+{
+	LocalVirtualWorkerId = InLocalVirtualWorkerId;
+	for (const auto& Elem : LayerNameToLBStrategy)
+	{
+		Elem.Value->SetLocalVirtualWorkerId(InLocalVirtualWorkerId);
+	}
+}
+
+bool ULayeredLBStrategy::ShouldHaveAuthority(const AActor& Actor) const
+{
+	if (!IsReady())
+	{
+		UE_LOG(LogLayeredLBStrategy, Warning, TEXT("LayeredLBStrategy not ready to relinquish authority for Actor %s."), *AActor::GetDebugName(&Actor));
+		return false;
+	}
+
+	const FName& LayerKey = GetLayerForActor(Actor);
+	if (!LayerNameToLBStrategy.Contains(LayerKey))
+	{
+		UE_LOG(LogLayeredLBStrategy, Warning, TEXT("LayeredLBStrategy doesn't have a LBStrategy for Actor %s which is in Layer %s."), *AActor::GetDebugName(&Actor), *LayerKey.ToString());
+		return false;
+	}
+
+	// If this worker is not responsible for the Actor's layer, just return false.
+	if (VirtualWorkerIdToLayerName.Contains(LocalVirtualWorkerId) && VirtualWorkerIdToLayerName[LocalVirtualWorkerId] != GetLayerForActor(Actor))
+	{
+		return false;
+	}
+
+	return LayerNameToLBStrategy[LayerKey]->ShouldHaveAuthority(Actor);
+}
+
+VirtualWorkerId ULayeredLBStrategy::WhoShouldHaveAuthority(const AActor& Actor) const
+{
+	if (!IsReady())
+	{
+		UE_LOG(LogLayeredLBStrategy, Warning, TEXT("LayeredLBStrategy not ready to decide on authority for Actor %s."), *AActor::GetDebugName(&Actor));
+		return SpatialConstants::INVALID_VIRTUAL_WORKER_ID;
+	}
+
+	const FName& LayerKey = GetLayerForActor(Actor);
```

Comment:
We seem to more than once try to get the LBStrategy for an actor. We have a function to get the LayorName for a given actor, perhaps we also might find use for a function that returns a LBStrategy* for a given actor?

Reply:
This would remove the repeated log: "LayeredLBStrategy doesn't have a LBStrategy for Actor %s which is in Layer %s"